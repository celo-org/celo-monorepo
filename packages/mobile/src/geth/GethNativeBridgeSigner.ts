import {
  encodeTransaction,
  extractSignature,
  RLPEncodedTx,
  rlpEncodedTx,
} from '@celo/contractkit/lib/utils/signing-utils'
import { Signer } from '@celo/contractkit/lib/wallets/signers/signer'
import { ensureLeading0x } from '@celo/utils/lib/address'
import { EIP712TypedData, generateTypedDataHash } from '@celo/utils/lib/sign-typed-data-utils'
import { normalizeAddressWith0x } from '@celo/utils/src/address'
import * as ethUtil from 'ethereumjs-util'
import { GethNativeModule } from 'react-native-geth'
import Logger from 'src/utils/Logger'
import { Tx } from 'web3-core'

const INCORRECT_PASSWORD_ERROR = 'could not decrypt key with given password'
const currentTimeInSeconds = () => Math.floor(Date.now() / 1000)

const TAG = 'geth/GethNativeBridgeSigner'
/**
 * Implements the signer interface on top of a RNGeth instance
 */
export class GethNativeBridgeSigner implements Signer {
  /**
   * Construct a new instance of the RPC signer
   *
   * @param geth the RNGeth instance
   * @param account Account address derived from the private key to be called in init
   * @param unlockBufferSeconds Number of seconds to shrink the unlocked duration by to account for
   * latency and timing inconsistencies on the node
   * @param unlockTime Timestamp in seconds when the signer was last unlocked
   * @param unlockDuration Number of seconds that the signer was last unlocked for
   */
  constructor(
    protected geth: GethNativeModule,
    protected account: string,
    protected unlockBufferSeconds = 5,
    protected unlockTime?: number,
    protected unlockDuration?: number
  ) {}

  async init(privateKey: string, passphrase: string) {
    return this.geth.addAccount(this.hexToBase64(privateKey), passphrase)
  }

  async signRawTransaction(tx: Tx) {
    if (normalizeAddressWith0x(tx.from! as string) !== this.account) {
      throw new Error(`RNGethSigner(${this.account}) cannot sign tx with 'from' ${tx.from}`)
    }
    const encodedTx = rlpEncodedTx(tx)
    const signature = await this.signTransaction(0, encodedTx)
    return encodeTransaction(encodedTx, signature)
  }

  async signTransaction(
    // addToV (chainId) is ignored here because geth will
    // build it based on its configuration
    addToV: number,
    encodedTx: RLPEncodedTx
  ): Promise<{ v: number; r: Buffer; s: Buffer }> {
    const signedTxBase64 = await this.geth.signTransaction(
      this.hexToBase64(encodedTx.rlpEncode),
      this.account
    )
    return extractSignature(this.base64ToHex(signedTxBase64))
  }

  async signPersonalMessage(data: string): Promise<{ v: number; r: Buffer; s: Buffer }> {
    Logger.info(`${TAG}@signPersonalMessage`, `Signing ${data}`)
    const hash = ethUtil.hashPersonalMessage(Buffer.from(data.replace('0x', ''), 'hex'))
    const signatureBase64 = await this.geth.signHash(hash.toString('base64'), this.account)
    return ethUtil.fromRpcSig(this.base64ToHex(signatureBase64))
  }

  async signTypedData(typedData: EIP712TypedData): Promise<{ v: number; r: Buffer; s: Buffer }> {
    // TODO: Not sure if it makes more sense to expose a `signTypedData` function on the RN Bridge
    // or just construct the hash here.
    Logger.info(`${TAG}@signTypedData`, `Signing typed data`)
    const hash = generateTypedDataHash(typedData)
    const signatureBase64 = await this.geth.signHash(hash.toString('base64'), this.account)
    return ethUtil.fromRpcSig(this.base64ToHex(signatureBase64))
  }

  getNativeKey = () => this.account

  async unlock(passphrase: string, duration: number): Promise<boolean> {
    try {
      // Duration in geth should be nanoseconds
      const durationNano = duration * 1000000000
      await this.geth.unlockAccount(this.account, passphrase, durationNano)
    } catch (error) {
      if (error?.message?.toLowerCase()?.includes(INCORRECT_PASSWORD_ERROR)) {
        return false
      }

      // Re-throw otherwise
      throw error
    }

    this.unlockTime = currentTimeInSeconds()
    this.unlockDuration = duration
    return true
  }

  isUnlocked(): boolean {
    if (this.unlockDuration === undefined || this.unlockTime === undefined) {
      return false
    }
    return this.unlockTime + this.unlockDuration - this.unlockBufferSeconds > currentTimeInSeconds()
  }

  decrypt(_ciphertext: Buffer) {
    return Promise.reject(new Error('Decryption operation is not supported on this signer'))
  }

  computeSharedSecret(_publicKey: string): Promise<Buffer> {
    return Promise.reject(new Error('computeSharedSecret is not supported on this signer'))
  }

  hexToBase64(hex: string) {
    return Buffer.from(hex.replace('0x', ''), 'hex').toString('base64')
  }

  base64ToHex(base64: string) {
    return ensureLeading0x(Buffer.from(base64, 'base64').toString('hex'))
  }
}
