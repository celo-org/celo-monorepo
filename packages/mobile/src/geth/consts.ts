export const WEI_PER_CELO = 1000000000000000000.0
export const UNLOCK_DURATION = 600

export enum GethSyncMode {
  // Valid sync mode values can be seen at https://github.com/celo-org/celo-blockchain/blob/8be27f7c044e35dbf63a42500a79805f1bddcfb8/mobile/geth.go#L43-L47
  // Anything invalid will cause Geth to panic and app to crash.
  Light = 3,
  // Value of 4 corresponds to a deprecated sync mode.
  Ultralight = 5,
}

// Re-export from utils for convinience since we use these often
export {
  CURRENCIES,
  currencyToShortMap,
  CURRENCY_ENUM,
  resolveCurrency,
  SHORT_CURRENCIES,
} from '@celo/utils/src/currencies'
