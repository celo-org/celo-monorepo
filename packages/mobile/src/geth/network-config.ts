import { DEFAULT_SYNC_MODE, Testnets } from 'src/web3/testnets'

export default {
  [Testnets.integration]: {
    nodeDir: `.${Testnets.integration}`,
    syncMode: DEFAULT_SYNC_MODE,
    blockchainApiUrl: 'https://integration-dot-celo-testnet.appspot.com/',
  },
  [Testnets.alfajoresstaging]: {
    nodeDir: `.${Testnets.alfajoresstaging}`,
    syncMode: DEFAULT_SYNC_MODE,
    blockchainApiUrl: 'https://alfajoresstaging-dot-celo-testnet.appspot.com/',
  },
  [Testnets.alfajores]: {
    nodeDir: `.${Testnets.alfajores}`,
    syncMode: DEFAULT_SYNC_MODE,
    blockchainApiUrl: 'https://alfajores-dot-celo-testnet-production.appspot.com/',
  },
  [Testnets.pilot]: {
    nodeDir: `.${Testnets.pilot}`,
    syncMode: DEFAULT_SYNC_MODE,
    blockchainApiUrl: 'https://pilot-dot-celo-testnet-production.appspot.com/',
  },
  [Testnets.pilotstaging]: {
    nodeDir: `.${Testnets.pilotstaging}`,
    syncMode: DEFAULT_SYNC_MODE,
    blockchainApiUrl: 'https://pilotstaging-dot-celo-testnet.appspot.com/',
  },
}
