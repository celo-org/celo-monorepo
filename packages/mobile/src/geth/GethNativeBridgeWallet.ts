import { CeloTx } from '@celo/connect'
import {
  ensureLeading0x,
  normalizeAddressWith0x,
  privateKeyToAddress,
} from '@celo/utils/lib/address'
import { UnlockableWallet } from '@celo/wallet-base'
import { RemoteWallet } from '@celo/wallet-remote'
import { GethNativeModule } from 'react-native-geth'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { GethNativeBridgeSigner } from 'src/geth/GethNativeBridgeSigner'
import Logger from 'src/utils/Logger'

const TAG = 'geth/GethNativeBridgeWallet'

export class GethNativeBridgeWallet extends RemoteWallet<GethNativeBridgeSigner>
  implements UnlockableWallet {
  /**
   * Construct a React Native geth wallet which uses the bridge methods
   * instead of communicating with a node
   * @param geth The instance of the bridge object
   * @dev overrides WalletBase.signTransaction
   */
  constructor(private geth: GethNativeModule) {
    super()
  }

  async loadAccountSigners(): Promise<Map<string, GethNativeBridgeSigner>> {
    let accounts: string[]
    const addressToSigner = new Map<string, GethNativeBridgeSigner>()

    try {
      accounts = await this.geth.listAccounts()
    } catch (e) {
      Logger.error(`${TAG}@loadAccountSigners`, 'Error listing accounts', e)
      throw new Error(ErrorMessages.GETH_FETCH_ACCOUNTS)
    }

    accounts.forEach((address) => {
      const cleanAddress = normalizeAddressWith0x(address)
      addressToSigner.set(cleanAddress, new GethNativeBridgeSigner(this.geth, cleanAddress))
    })
    return addressToSigner
  }

  async addAccount(privateKey: string, passphrase: string): Promise<string> {
    Logger.info(`${TAG}@addAccount`, `Adding a new account`)
    const address = normalizeAddressWith0x(privateKeyToAddress(ensureLeading0x(privateKey)))
    if (this.hasAccount(address)) {
      throw new Error(ErrorMessages.GETH_ACCOUNT_ALREADY_EXISTS)
    }
    const signer = new GethNativeBridgeSigner(this.geth, address)
    const resultantAddress = await signer.init(privateKey, passphrase)
    if (normalizeAddressWith0x(resultantAddress) !== address) {
      throw new Error(ErrorMessages.GETH_UNEXPECTED_ADDRESS_ON_ADD)
    }
    this.addSigner(resultantAddress, signer)
    return resultantAddress
  }

  /**
   * Unlocks an account for a given duration
   * @param account String the account to unlock
   * @param passphrase String the passphrase of the account
   * @param duration Number the duration of the unlock period
   */
  async unlockAccount(account: string, passphrase: string, duration: number) {
    Logger.info(`${TAG}@unlockAccount`, `Unlocking ${account}`)
    const signer = this.getSigner(account) as GethNativeBridgeSigner
    return signer.unlock(passphrase, duration)
  }

  isAccountUnlocked(address: string) {
    const signer = this.getSigner(address)
    return signer.isUnlocked()
  }

  /**
   * Gets the signer based on the 'from' field in the tx body
   * @param txParams Transaction to sign
   * @dev overrides WalletBase.signTransaction
   */
  async signTransaction(txParams: CeloTx) {
    Logger.info(`${TAG}@signTransaction`, `Signing transaction: ${JSON.stringify(txParams)}`)
    // Get the signer from the 'from' field
    const fromAddress = txParams.from!.toString()
    const signer = this.getSigner(fromAddress)
    return signer.signRawTransaction(txParams)
  }
}
