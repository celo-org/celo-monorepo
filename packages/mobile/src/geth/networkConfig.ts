import { DEFAULT_SYNC_MODE, DEFAULT_TESTNET, FORNO_ENABLED_INITIALLY } from 'src/config'
import { GethSyncMode } from 'src/geth/consts'

export enum Testnets {
  integration = 'integration',
  alfajoresstaging = 'alfajoresstaging',
  alfajores = 'alfajores',
  pilot = 'pilot',
  pilotstaging = 'pilotstaging',
  baklavastaging = 'baklavastaging',
  baklava = 'baklava',
}

interface NetworkConfig {
  nodeDir: string
  syncMode: GethSyncMode
  initiallyForno: boolean
  blockchainApiUrl: string
}

const networkConfigs: { [testnet: string]: NetworkConfig } = {
  [Testnets.integration]: {
    nodeDir: `.${Testnets.integration}`,
    syncMode: DEFAULT_SYNC_MODE,
    initiallyForno: FORNO_ENABLED_INITIALLY,
    blockchainApiUrl: 'https://integration-dot-celo-testnet.appspot.com/',
  },
  [Testnets.alfajoresstaging]: {
    nodeDir: `.${Testnets.alfajoresstaging}`,
    syncMode: DEFAULT_SYNC_MODE,
    initiallyForno: FORNO_ENABLED_INITIALLY,
    blockchainApiUrl: 'https://alfajoresstaging-dot-celo-testnet.appspot.com/',
  },
  [Testnets.alfajores]: {
    nodeDir: `.${Testnets.alfajores}`,
    syncMode: DEFAULT_SYNC_MODE,
    initiallyForno: FORNO_ENABLED_INITIALLY,
    blockchainApiUrl: 'https://alfajores-dot-celo-testnet-production.appspot.com/',
  },
  [Testnets.pilot]: {
    nodeDir: `.${Testnets.pilot}`,
    syncMode: DEFAULT_SYNC_MODE,
    initiallyForno: FORNO_ENABLED_INITIALLY,
    blockchainApiUrl: 'https://pilot-dot-celo-testnet-production.appspot.com/',
  },
  [Testnets.pilotstaging]: {
    nodeDir: `.${Testnets.pilotstaging}`,
    syncMode: DEFAULT_SYNC_MODE,
    initiallyForno: FORNO_ENABLED_INITIALLY,
    blockchainApiUrl: 'https://pilotstaging-dot-celo-testnet.appspot.com/',
  },
  [Testnets.baklavastaging]: {
    nodeDir: `.${Testnets.baklavastaging}`,
    syncMode: DEFAULT_SYNC_MODE,
    initiallyForno: FORNO_ENABLED_INITIALLY,
    blockchainApiUrl: 'https://baklavastaging-dot-celo-testnet.appspot.com/',
  },
  [Testnets.baklava]: {
    nodeDir: `.${Testnets.baklava}`,
    syncMode: DEFAULT_SYNC_MODE,
    initiallyForno: FORNO_ENABLED_INITIALLY,
    blockchainApiUrl: 'https://baklava-dot-celo-testnet-production.appspot.com/',
  },
}

export default networkConfigs[DEFAULT_TESTNET]
