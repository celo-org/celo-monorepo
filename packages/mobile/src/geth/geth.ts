import { GenesisBlockUtils, StaticNodeUtils } from '@celo/walletkit'
import BigNumber from 'bignumber.js'
import { Platform } from 'react-native'
import DeviceInfo from 'react-native-device-info'
import * as RNFS from 'react-native-fs'
import RNGeth from 'react-native-geth'
import { DEFAULT_TESTNET } from 'src/config'
import networkConfig from 'src/geth/networkConfig'
import Logger from 'src/utils/Logger'
import FirebaseLogUploader from 'src/utils/LogUploader'

let gethLock = false
let gethInstance: typeof RNGeth | null = null

export const FailedToFetchStaticNodesError = new Error(
  'Failed to fetch static nodes from Google storage'
)
export const FailedToFetchGenesisBlockError = new Error(
  'Failed to fetch genesis block from Google storage'
)

// We are never going to run mobile node in full or fast mode.
enum SyncMode {
  LIGHT = 'light',
  ULTRA_LIGHT = 'ultralight',
}

// Log levels correpond to the values defined in
// https://github.com/celo-org/geth/blob/master/log/logger.go#L21
enum LogLevel {
  CRITICAL = 0,
  ERROR = 1,
  WARN = 2,
  INFO = 3,
  DEBUG = 4,
  TRACE = 5,
}

// The logs will be uploaded only if they are larger than this size
const UPLOAD_SIZE_THRESHOLD = 10 * 1024 // 10 KB

enum ErrorType {
  Unknown,
  GethAlreadyRunning,
  CorruptChainData,
}

// Must match `clientIdentifier`
// see https://github.com/celo-org/celo-blockchain/blob/d4b48f3e79b01e8cb7dcf8606b0ed1f666a37a2f/mobile/geth.go#L143
// and https://github.com/celo-org/celo-blockchain/blob/d4b48f3e79b01e8cb7dcf8606b0ed1f666a37a2f/mobile/geth_android.go
const INSTANCE_FOLDER = Platform.select({
  ios: 'iGeth',
  android: 'GethDroid',
  default: 'GethMobile',
})

function getNodeInstancePath(nodeDir: string) {
  return `${RNFS.DocumentDirectoryPath}/${nodeDir}/${INSTANCE_FOLDER}`
}

function getFolder(filePath: string) {
  return filePath.substr(0, filePath.lastIndexOf('/'))
}

async function createNewGeth(): Promise<typeof RNGeth> {
  Logger.debug('Geth@newGeth', 'Configure and create new Geth')
  const { nodeDir, syncMode } = networkConfig
  const genesis: string = await readGenesisBlockFile(nodeDir)
  const networkID: number = GenesisBlockUtils.getChainIdFromGenesis(genesis)

  Logger.debug('Geth@newGeth', `Network ID is ${networkID}`)

  const gethOptions: any = {
    nodeDir,
    networkID,
    genesis,
    syncMode,
    useLightweightKDF: true,
  }

  // Setup Logging
  const logFilePath = Logger.getGethLogFilePath()

  // Upload logs first
  await uploadLogs(logFilePath, Logger.getReactNativeLogsFilePath())
  gethOptions.logFile = logFilePath
  // Only log info and above to the log file.
  // The logcat logging mode remains unchanged.
  gethOptions.logFileLogLevel = LogLevel.INFO
  Logger.debug('Geth@newGeth', 'Geth logs will be piped to ' + logFilePath)

  return new RNGeth(gethOptions)
}

async function initGeth() {
  Logger.info('Geth@init', 'Create a new Geth instance')

  if (gethLock) {
    Logger.warn('Geth@init', 'Geth create already in progress.')
    return
  }
  gethLock = true

  try {
    if (gethInstance) {
      Logger.debug('Geth@init', 'Geth already exists, trying to stop it.')
      await stop()
    }

    if (!(await ensureGenesisBlockWritten())) {
      throw FailedToFetchGenesisBlockError
    }
    if (!(await ensureStaticNodesInitialized())) {
      throw FailedToFetchStaticNodesError
    }
    const geth = await createNewGeth()

    try {
      await geth.start()
      gethInstance = geth
      geth.subscribeNewHead()
    } catch (e) {
      const errorType = getGethErrorType(e)
      if (errorType === ErrorType.GethAlreadyRunning) {
        // Geth is already running, this is most likely RN restart.
        Logger.info('Geth@init/startInstance', 'Geth start reported geth already running')
        // Note: Unfortunately, RN-Geth doesn't currently support connecting to the
        // already running geth instance, which we would need to subscribe to head updates.
        // In the meantime, we need to force an app reset. See #3227
        throw new Error('Geth already running, need to restart app')
      } else if (errorType === ErrorType.CorruptChainData) {
        Logger.warn('Geth@init/startInstance', 'Geth start reported chain data error')
        await attemptGethCorruptionFix(geth)
      } else {
        Logger.error('Geth@init/startInstance', 'Unexpected error starting geth', e)
        throw e
      }
    }
  } finally {
    gethLock = false
  }
}

export async function getGeth(): Promise<typeof gethInstance> {
  Logger.debug('Geth@getGeth', 'Getting Geth Instance')
  if (!gethInstance) {
    await initGeth()
  }
  return gethInstance
}

async function ensureStaticNodesInitialized(): Promise<boolean> {
  const { nodeDir } = networkConfig
  if (await staticNodesAlreadyInitialized(nodeDir)) {
    Logger.debug('Geth@maybeInitStaticNodes', 'static nodes already initialized')
    return true
  } else {
    Logger.debug('Geth@maybeInitStaticNodes', 'initializing static nodes')
    let enodes: string | null = null
    try {
      enodes = await StaticNodeUtils.getStaticNodesAsync(DEFAULT_TESTNET)
    } catch (error) {
      Logger.error(
        `Failed to get static nodes for network ${DEFAULT_TESTNET},` +
          `the node will not be able to sync with the network till restart`,
        error
      )
      return false
    }
    if (enodes != null) {
      await writeStaticNodes(nodeDir, enodes)
      return true
    }
    return false
  }
}

async function stop() {
  try {
    Logger.debug('Geth@stop', 'Stopping Geth')
    await gethInstance.stop()
    Logger.debug('Geth@stop', 'Geth stopped')
  } catch (e) {
    Logger.error('Geth@stop', 'Error stopping Geth', e)
    throw e
  }
}

async function ensureGenesisBlockWritten(): Promise<boolean> {
  const { nodeDir } = networkConfig
  if (await genesisBlockAlreadyWritten(nodeDir)) {
    Logger.debug('Geth@ensureGenesisBlockWritten', 'genesis block already written')
    return true
  } else {
    Logger.debug('Geth@ensureGenesisBlockWritten', 'writing genesis block')
    let genesisBlock: string | null = null
    try {
      genesisBlock = await GenesisBlockUtils.getGenesisBlockAsync(DEFAULT_TESTNET)
    } catch (error) {
      Logger.error(`Failed to get the genesis block for network ${DEFAULT_TESTNET}.`, error)
      return false
    }
    if (genesisBlock != null) {
      await writeGenesisBlock(nodeDir, genesisBlock)
      return true
    }
    return false
  }
}

function getGenesisBlockFile(nodeDir: string) {
  return `${getNodeInstancePath(nodeDir)}/genesis.json`
}

async function genesisBlockAlreadyWritten(nodeDir: string): Promise<boolean> {
  const genesisBlockFile = getGenesisBlockFile(nodeDir)
  if (!(await RNFS.exists(genesisBlockFile))) {
    return false
  }
  const fileStat: RNFS.StatResult = await RNFS.stat(genesisBlockFile)
  return fileStat.isFile() && new BigNumber(fileStat.size, 10).isGreaterThan(0)
}

async function readGenesisBlockFile(nodeDir: string): Promise<string> {
  const genesisBlockFile = getGenesisBlockFile(nodeDir)
  return RNFS.readFile(genesisBlockFile, { encoding: 'utf8' })
}

async function writeGenesisBlock(nodeDir: string, genesisBlock: string) {
  Logger.debug(`writeGenesisBlock genesis block is: "${genesisBlock}"`)
  const genesisBlockFile = getGenesisBlockFile(nodeDir)
  await RNFS.mkdir(getFolder(genesisBlockFile))
  await RNFS.writeFile(genesisBlockFile, genesisBlock, 'utf8')
}

function getStaticNodesFile(nodeDir: string) {
  return `${getNodeInstancePath(nodeDir)}/static-nodes.json`
}

/**
 * Returns true if the static nodes files in the Geth data dir has been initialized, false otherwise.
 * @param nodeDir Geth data dir
 */
async function staticNodesAlreadyInitialized(nodeDir: string): Promise<boolean> {
  const staticNodesFile = getStaticNodesFile(nodeDir)
  if (!(await RNFS.exists(staticNodesFile))) {
    return false
  }
  const fileStat: RNFS.StatResult = await RNFS.stat(staticNodesFile)
  return fileStat.isFile() && new BigNumber(fileStat.size, 10).isGreaterThan(0)
}

async function writeStaticNodes(nodeDir: string, enodes: string) {
  console.info(`writeStaticNodes enodes are "${enodes}"`)
  const staticNodesFile = getStaticNodesFile(nodeDir)
  await RNFS.mkdir(getFolder(staticNodesFile))
  await RNFS.writeFile(staticNodesFile, enodes, 'utf8')
}

async function attemptGethCorruptionFix(geth: any) {
  const deleteChainDataResult = await deleteChainData()
  const deleteGethLockResult = await deleteGethLockFile()
  if (deleteChainDataResult && deleteGethLockResult) {
    await geth.start()
    gethInstance = geth
    geth.subscribeNewHead()
  } else {
    throw new Error('Failed to fix Geth and restart')
  }
}

export async function deleteChainData() {
  Logger.debug('Geth@deleteChainData', 'Deleting chain data')
  // Delete data for both the possible modes a mobile node could be running in.
  const result1 = await deleteSingleChainData(SyncMode.ULTRA_LIGHT)
  const result2 = await deleteSingleChainData(SyncMode.LIGHT)
  return result1 || result2
}

async function deleteSingleChainData(syncMode: SyncMode) {
  const { nodeDir } = networkConfig
  const chainDataDir = `${getNodeInstancePath(nodeDir)}/${syncMode}chaindata`
  Logger.debug('Geth@deleteSingleChainData', `Going to delete ${chainDataDir}`)
  return deleteFileIfExists(chainDataDir)
}

async function deleteGethLockFile() {
  // Delete the .ipc file or the Geth will think that some other Geth node is using this datadir.
  const { nodeDir } = networkConfig
  const gethLockFile = `${getNodeInstancePath(nodeDir)}/LOCK`
  Logger.info('Geth@deleteGethLockFile', `Deleting ${gethLockFile} for nodeDir ${nodeDir}`)
  return deleteFileIfExists(gethLockFile)
}

async function deleteFileIfExists(path: string) {
  try {
    const gethLockFileExists = await RNFS.exists(path)
    if (gethLockFileExists) {
      Logger.debug('Geth@deleteFileIfExists', `Dir ${path} exists. Attempting to delete`)
      await RNFS.unlink(path)
      return true
    } else {
      Logger.debug('Geth@deleteFileIfExists', `Dir ${path} does not exist`)
      return true
    }
  } catch (error) {
    Logger.error('Geth@deleteFileIfExists', `Failed to delete ${path}`, error)
    return false
  }
}

// The only reason to upload both the logs simulatenously here is to have the same upload ID for both, so that,
// the developers can correlate them.
async function uploadLogs(gethLogFilePath: string, reactNativeLogFilePath: string) {
  Logger.debug('Geth@uploadLogs', 'Attempting to upload geth logs')
  try {
    const bundleId = DeviceInfo.getBundleId()
    const uploadPath = `${bundleId}/${DEFAULT_TESTNET}`

    // Phone number might not be verified here but that does not matter for logging.
    const phoneNumber = (await DeviceInfo.getPhoneNumber()) || 'unknown'
    const timestamp = new Date().getTime()
    const deviceId = DeviceInfo.getUniqueID()
    const uploadId = `${timestamp}_${deviceId}`
    const gethUploadFileName = `${phoneNumber}_${uploadId}_geth.txt`
    const reactNativeUploadFileName = `${phoneNumber}_${uploadId}_rn.txt`
    // Upload one if the other one is uploaded.

    const [shouldUploadGeth, shouldUploadRN] = await Promise.all([
      FirebaseLogUploader.shouldUpload(gethLogFilePath, UPLOAD_SIZE_THRESHOLD, true),
      FirebaseLogUploader.shouldUpload(reactNativeLogFilePath, UPLOAD_SIZE_THRESHOLD, true),
    ])

    // If either of them have to be uploaded then upload both.
    // Noth that the Wi-Fi can switch to cellular between the time of check and
    // the time of use but at this time that's an acceptable tradeoff.
    if (shouldUploadGeth || shouldUploadRN) {
      await Promise.all([
        FirebaseLogUploader.upload(gethLogFilePath, uploadPath, gethUploadFileName),
        FirebaseLogUploader.upload(reactNativeLogFilePath, uploadPath, reactNativeUploadFileName),
      ])
    }
  } catch (e) {
    Logger.error('Geth@uploadLogs', 'Failed to upload logs', e)
  }
}

function getGethErrorType(e: Error): ErrorType {
  if (!e || !e.message) {
    return ErrorType.Unknown
  }
  if (e.message.includes('datadir already used by another process')) {
    return ErrorType.GethAlreadyRunning
  }
  if (e.message.includes('missing block number for head header hash')) {
    return ErrorType.CorruptChainData
  }
  return ErrorType.Unknown
}
