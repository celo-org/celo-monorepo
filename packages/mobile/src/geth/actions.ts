import { InitializationState } from 'src/geth/reducer'

export enum Actions {
  SET_INIT_STATE = 'GETH/SET_INIT_STATE',
  SET_GETH_CONNECTED = 'GETH/SET_GETH_CONNECTED',
  CANCEL_GETH_SAGA = 'GETH/CANCEL_GETH_SAGA',
}

interface SetInitStateAction {
  type: Actions.SET_INIT_STATE
  state: InitializationState
}

export const setInitState = (state: InitializationState): SetInitStateAction => ({
  type: Actions.SET_INIT_STATE,
  state,
})

export const cancelGethSaga = () => ({
  type: Actions.CANCEL_GETH_SAGA,
})

interface SetGethConnectedAction {
  type: Actions.SET_GETH_CONNECTED
  connected: boolean
}

export const setGethConnected = (connected: boolean): SetGethConnectedAction => ({
  type: Actions.SET_GETH_CONNECTED,
  connected,
})

export type ActionTypes = SetInitStateAction | SetGethConnectedAction
