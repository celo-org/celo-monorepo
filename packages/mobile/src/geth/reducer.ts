import { Actions, ActionTypes } from 'src/geth/actions'

export enum InitializationState {
  NOT_YET_INITIALIZED = 'NOT_YET_INITIALIZED',
  INITIALIZING = 'INITIALIZING',
  INITIALIZED = 'INITIALIZED',
  INITIALIZE_ERROR = 'INITIALIZE_ERROR',
  DATA_CONNECTION_MISSING_ERROR = 'DATA_CONNECTION_MISSING_ERROR',
}

export interface State {
  initialized: InitializationState
  connected: boolean
  promptZeroSyncIfNeeded: boolean
}

const initialState: State = {
  initialized: InitializationState.NOT_YET_INITIALIZED,
  connected: false,
  promptZeroSyncIfNeeded: false,
}

export function gethReducer(state: State = initialState, action: ActionTypes) {
  switch (action.type) {
    case Actions.SET_INIT_STATE:
      return { ...state, initialized: action.state }
    case Actions.SET_GETH_CONNECTED:
      return {
        ...state,
        connected: action.connected,
      }
    case Actions.SET_PROMPT_ZERO_SYNC:
      return {
        ...state,
        promptZeroSyncIfNeeded: action.promptIfNeeded,
      }

    default:
      return state
  }
}
