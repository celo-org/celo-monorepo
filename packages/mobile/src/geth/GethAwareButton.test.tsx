import { BtnTypes } from '@celo/react-components/components/Button'
import * as React from 'react'
import 'react-native'
import * as renderer from 'react-test-renderer'
import { GethAwareButton } from 'src/geth/GethAwareButton'

it('renders correctly when disconnected', () => {
  const tree = renderer.create(
    <GethAwareButton
      onPress={jest.fn()}
      connected={false}
      text={'Celo Button'}
      standard={false}
      type={BtnTypes.TERTIARY}
    />
  )
  expect(tree).toMatchSnapshot()
})

it('renders correctly when connected', () => {
  const tree = renderer.create(
    <GethAwareButton
      onPress={jest.fn()}
      connected={true}
      text={'Celo Button'}
      standard={false}
      type={BtnTypes.TERTIARY}
    />
  )
  expect(tree).toMatchSnapshot()
})
