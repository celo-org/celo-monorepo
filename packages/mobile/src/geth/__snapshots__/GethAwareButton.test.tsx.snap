// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when connected 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 3,
        "flexDirection": "row",
      },
      null,
      null,
      undefined,
      Object {
        "backgroundColor": "transparent",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    nativeBackgroundAndroid={
      Object {
        "attribute": "selectableItemBackground",
        "type": "ThemeAttrAndroid",
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "transparent",
        },
        Object {
          "height": 50,
        },
        Object {
          "borderWidth": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Hind-SemiBold",
              "fontSize": 16,
            },
            Object {
              "color": "#2E3338",
            },
            Object {
              "paddingLeft": 5,
              "paddingRight": 5,
            },
          ]
        }
      >
        Celo Button
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders correctly when disconnected 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 3,
        "flexDirection": "row",
      },
      null,
      null,
      undefined,
      Object {
        "backgroundColor": "transparent",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    nativeBackgroundAndroid={
      Object {
        "attribute": "selectableItemBackground",
        "type": "ThemeAttrAndroid",
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "backgroundColor": "transparent",
        },
        Object {
          "height": 50,
        },
        Object {
          "borderWidth": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Hind-SemiBold",
              "fontSize": 16,
            },
            Object {
              "color": "color: rgba(46, 51, 56, 0.4)",
            },
            Object {
              "paddingLeft": 5,
              "paddingRight": 5,
            },
          ]
        }
      >
        Celo Button
      </Text>
    </View>
  </View>
</View>
`;
